### PR TITLE
Trim unused CSS styles and minify in staging / prod

### PR DESCRIPTION
**Pull request by syntapy**
### Description checklist

Does several things
- Upgrades tailwindcss from v2 -> v3, which also results in stronger borders during tab traversal (possibly helps #1906  but I can't open those docs to verify)
- Uses regex 'parsing' to reduce the 1.7M `tailwind.css` file in v2 down to (in v3) 33K in dev (unminified) and 23K in staging & prod (minified)
- Reducing tailwind.css in v3 also reduces time to run `bin/refresh-styles` down to 1s or 2s, from about 8s in v2 without trimming styles if I remember correctly
- Remaining task is to use SBT to automatically recompile `tailwind.css` whenever any of the UI code in `app/views/**/*.java` is modified. As such, devs should not need to worry about manually running `bin/refresh-styles`

I only have Linux / Firefox / Chromium, so it would be helpful if people on Mac / Windows can check the styles

Reduction in `tailwind.css` might help poor people the most since they might have limited data, and I'm guessing probably benefit the most from this site

> If ye meet the abased or the down-trodden, turn not away disdainfully from them, for the King of Glory ever watcheth over them and surroundeth them with such tenderness as none can fathom... -Baha'u'llah

### Checklist
- [x] Regex parsing
- [x] Upgrade tailwindcss from v2 -> v3
  - [x] Give up on trying to elliminate final conversion warning
- [x] Check styles in firefox / chrome device emulators
  - [ ] Investigate tab traverse borders red in chromium but blue in firefox 
- [x] Do shanemc's comments
  - [x] Exclude `Styles.java` and `ReferenceClasses.java` from `bin/fmt`
- [ ] Automate tailwind.css recompilation with sbt (I tried with Grunt but that was too flakey)
- [x] Check with team how it relates to #1906 
- [x] Update styles README on code constraints

### Issues

Fixes https://github.com/seattle-uat/civiform/issues/1940
Fixes https://github.com/seattle-uat/civiform/issues/1480
Partially addresses https://github.com/seattle-uat/civiform/issues/1906

Related to https://github.com/seattle-uat/civiform/issues/1399

### Related PRs
https://github.com/seattle-uat/civiform/pull/1386
https://github.com/seattle-uat/civiform/pull/1392/
https://github.com/seattle-uat/civiform/pull/1419

### CODE REVIEW

File: universal-application-tool-0.0.1/tags.txt
Comment (syntapy) [inline universal-application-tool-0.0.1/tags.txt:1]: This is a list of all the html tags. I had to add them for some reason to make it work

Comment (syntapy) [reply universal-application-tool-0.0.1/tags.txt:1]: I think I just realized why I had to do this. I guess tailwind parses which html tags are being used when creating its .css, so I need to parse which `j2html` tags are being used. This would probably bring down the size a bit from where it is now (155K)

File: universal-application-tool-0.0.1/tailwind.config.js
Comment (bion) [inline universal-application-tool-0.0.1/tailwind.config.js:21]: suggestion (blocking): extract the regexes to constants

File: universal-application-tool-0.0.1/tailwind.config.js
Comment (bion) [inline universal-application-tool-0.0.1/tailwind.config.js:37]: suggestion (blocking): throw an error instead

File: universal-application-tool-0.0.1/tailwind.config.js
Comment (bion) [inline universal-application-tool-0.0.1/tailwind.config.js:41]:
```suggestion
      count++
```

File: bin/run-dev
Comment (bion) [inline bin/run-dev:43]: suggestion (blocking): please revert this change

File: bin/refresh-styles
Comment (bion) [inline bin/refresh-styles:11]: ah good catch, thank you

**Review by bion — CHANGES_REQUESTED** [review]
This looks a pretty workable approach! I added a few comments mostly on style, the PR description mentions doing a little more work on it so I'll leave it at just those comments for now.

Great work so far!

File: universal-application-tool-0.0.1/tailwind.config.js
Comment (syntapy) [reply universal-application-tool-0.0.1/tailwind.config.js:35]: Oh, I meant to use logical `&&` instead of bitwise lol. It was just print line for debugging

File: universal-application-tool-0.0.1/tailwind.config.js
Comment (bion) [inline universal-application-tool-0.0.1/tailwind.config.js:114]: issue (blocking): the transform is called "java" but also processes TS files

File: universal-application-tool-0.0.1/tailwind.config.js
Comment (bion) [inline universal-application-tool-0.0.1/tailwind.config.js:44]:
```suggestion
      let data = fs.readFileSync(folder + file, 'utf8').split('\n');
```

File: universal-application-tool-0.0.1/tailwind.config.js
Comment (bion) [inline universal-application-tool-0.0.1/tailwind.config.js:58]:
```suggestion
```

**Review by bion — CHANGES_REQUESTED** [review]

**Comment by bion** [comment]
@syntapy thanks! please merge in the `main` branch with this one to ensure it's up to date before merging

**Comment by bion** [comment]
@syntapy need to run `bin/fmt` to fix formatting

**Comment by bion** [comment]
@shanemc-goog @gwendolyngoetz @goldblatt please review when you've got time. Huge improvement in page weight but want to make sure the dev team is aware of and up for this change.

**Comment by bion** [comment]
@shanemc-goog @gwendolyngoetz @goldblatt please review when you've got time. Huge improvement in page weight but want to make sure the dev team is aware of and up for this change.

**Comment by syntapy** [comment]
@bion Just had a thought: It might be usefull to add an option to not trim styles in dev, so that you all dont have to constantly refresh after every change when experimenting with different css classes

**Comment by bion** [comment]
> It might be usefull to add an option to not trim styles in dev, so that you all dont have to constantly refresh after every change when experimenting with different css classes

@syntapy this is worth the extra complexity imo, reloading is pretty slow atm.

File: universal-application-tool-0.0.1/tailwind.config.js
Comment (shanemc-goog) [inline universal-application-tool-0.0.1/tailwind.config.js:49]: rename special as I think they're the "normal" files for this operation

styleFiles maybe?

Comment (shanemc-goog) [inline universal-application-tool-0.0.1/tailwind.config.js:49]: Suggestion: make this a file level variable next to the RGX that references them too, then their relationship is clearer and updating them will be less error prone.

File: universal-application-tool-0.0.1/tailwind.config.js
Comment (shanemc-goog) [inline universal-application-tool-0.0.1/tailwind.config.js:26]: Suggestion: rename to line_number, set to 0 and do ++line_number as the first line in the for.

**Comment by syntapy** [comment]
@bion  Fyi I'm working on upgrading from tailwind v2 -> v3. This increases resulting style file (so far) from 155K to 252K, but it decreases the time to run `bin/refresh-styles` from about 21s to like 5s or 6s on my machine. However, a couple styles are kinda off as a result so I'm debugging that. I guess this would remove the need to not trim in dev then?

@shanemc-goog Thanks for the comments. I will do them

**Comment by bion** [comment]
> I guess this would remove the need to not trim in dev then?

Is the proposed workflow for dev then the following?
1. edit a j2html template to use a previously unused tailwind style class
2. manually run `bin/refresh-styles`
3. reload page

If so I could see that being annoying for front-end devs having uncertainty about when they need to do step 2, with the only way of knowing if the reason you're not getting the desired result is to do step 2 every time, which adds another 5-6 seconds every time you make an edit.

**Comment by syntapy** [comment]
@bion Maybe theres an option to run `refresh-styles` on hot reloading compilations

I've got an option working to not minify in dev

But under tailwind v3 it results in 3.1M style file and  takes 38s to run on my machine. I guess that would only add to the beginning build time but its still pretty long

Granted my laptop is a bit old so the 38s & 5s would be shorter on a machine used by a typical developer on your team I imagine

It seems most of the computation time is actually looping through the list of all styles so I could try parallelizing

**Comment by syntapy** [comment]
I can't figure out why I'm still getting this upgrade warning 

```
warn - The `purge`/`content` options have changed in Tailwind CSS v3.0.
warn - Update your configuration file to eliminate this warning.
warn - https://tailwindcss.com/docs/upgrade-guide#configure-content-sources
```

I've went through the [upgrade guide](https://tailwindcss.com/docs/upgrade-guide) a couple times, even deleted all contents of the tailwind config file, and it still there. So I guess it will remain there for now

**Comment by syntapy** [comment]
I can probably use regex to shrink the style file even more by trimming out unused media query combinations. This will also speed up building the CSS style file by a factor of 2 or 3. Which basically solves the question of whether I should incorporate this into SBT build process

**Comment by syntapy** [comment]
Oh weird. On firefox mobile emulator borders are blue when tab traversing, but on chromium their red

**Comment by syntapy** [comment]
Oh weird. On firefox mobile emulator borders are blue when tab traversing, but on chromium their red

File: bin/build-and-push-formatter
Comment (bion) [inline bin/build-and-push-formatter:12]: issue (blocking): please revert

File: bin/refresh-styles
Comment (bion) [inline bin/refresh-styles:7]: suggestion (blocking): mention the `--minify` flag in the docstring on L3

File: bin/run-dev
Comment (bion) [inline bin/run-dev:60]: issue (blocking): please revert

**Review by bion — CHANGES_REQUESTED** [review]

**Review by bion — DISMISSED** [review]

File: bin/pull-image
Comment (goldblatt) [inline bin/pull-image:21]: i think this will need to get rebased with #2262

**Comment by syntapy** [comment]
Yeah the Grunt stuff is not running on the new rebase

File: universal-application-tool-0.0.1/Gruntfile.js
Comment (kmcnellis) [reply universal-application-tool-0.0.1/Gruntfile.js:10]: Alternately, it might be clearer to make grunt the only way to run tailwinds, and just keep all the logic on calling it in one place (here)

Comment (kmcnellis) [reply universal-application-tool-0.0.1/Gruntfile.js:10]: A totally optional alternative: looks like there's a way to do this just in SBT (using sbt-web), but there's minimal-to-no docs on it.  It would be cool, though!  Maybe something for the future

https://github.com/sbt/sbt-web

One example:
https://github.com/sbt/sbt-uglify/blob/02f0e99a58fc9cefb268278379f1bfce490a87e6/src/main/scala/com/typesafe/sbt/uglify/SbtUglify.scala

File: universal-application-tool-0.0.1/project/TailwindBuilder.scala
Comment (kmcnellis) [inline universal-application-tool-0.0.1/project/TailwindBuilder.scala:13]: Is this npx tailwinds still required here?

**Review by kmcnellis — CHANGES_REQUESTED** [review]

File: universal-application-tool-0.0.1/project/TailwindBuilder.scala
Comment (syntapy) [reply universal-application-tool-0.0.1/project/TailwindBuilder.scala:13]: The grunt processes only takes care of compiling on the changes. So either the npx tailwinds here or the dev one in the build.sbt is needed I'm guessing. I've mentioned that in another comment on this

File: universal-application-tool-0.0.1/project/TailwindBuilder.scala
Comment (syntapy) [reply universal-application-tool-0.0.1/project/TailwindBuilder.scala:20]: Yeah it might be better to use just plain sbt over grunt. It seems more elegant than adding another layer of code watching. I tried that at first but it was pretty confusing, but I'll take a look at these links you sent me and I will reach out for more help if I can't figure it out

Comment (syntapy) [reply universal-application-tool-0.0.1/project/TailwindBuilder.scala:20]: Thank you I appreciate this

File: universal-application-tool-0.0.1/app/views/style/Styles.java
Comment (shanemc-goog) [inline universal-application-tool-0.0.1/app/views/style/Styles.java:33]: this is missing a word I think.

File: universal-application-tool-0.0.1/tailwind.config.js
Comment (shanemc-goog) [inline universal-application-tool-0.0.1/tailwind.config.js:50]: These need to be camel cased.  We use the google style guides.

https://google.github.io/styleguide/jsguide.html#naming-rules-common-to-all-identifiers

File: universal-application-tool-0.0.1/tailwind.config.js
Comment (shanemc-goog) [inline universal-application-tool-0.0.1/tailwind.config.js:20]: Is this not const for a reason?

File: universal-application-tool-0.0.1/tailwind.config.js
Comment (shanemc-goog) [inline universal-application-tool-0.0.1/tailwind.config.js:26]: These should be ALL_CAP case I believe since they appear immutable.

**Review by shanemc-goog — CHANGES_REQUESTED** [review]

File: bin/refresh-styles
Comment (kmcnellis) [reply bin/refresh-styles]: Line 3

File: bin/pull-image
Comment (kmcnellis) [inline bin/pull-image:8]: Can you keep this line? "Making sure we're up to date with the latest dev..."

File: bin/pull-image
Comment (kmcnellis) [inline bin/pull-image:14]: And this one as well "Using local civiform docker images"

File: bin/pull-image
Comment (kmcnellis) [inline bin/pull-image:15]: It's slightly confusing that the instructions to show download progress are only show when they are already being used.  Perhaps make showing progress the default, and you can use a flag to skip.

File: bin/pull-image
Comment (kmcnellis) [inline bin/pull-image:14]: Add the PULL_PROGRESS flag to the docstring at the top of this file.

File: universal-application-tool-0.0.1/app/views/style/README.md
Comment (kmcnellis) [inline universal-application-tool-0.0.1/app/views/style/README.md:45]: Is it still true that you have to restart the server with the new grunt changes?

File: universal-application-tool-0.0.1/app/views/style/Styles.java
Comment (kmcnellis) [inline universal-application-tool-0.0.1/app/views/style/Styles.java:2516]: [suggestion only - not a blocker] In the future, I'd suggest makeing these css changes in a seperate PR than the regex stuff, just to make it easier to review

File: universal-application-tool-0.0.1/package-lock.json
Comment (kmcnellis) [inline universal-application-tool-0.0.1/package-lock.json:1]: [blocking] Delete this file

File: universal-application-tool-0.0.1/app/views/admin/programs/ProgramBlockEditView.java
Comment (syntapy) [reply universal-application-tool-0.0.1/app/views/admin/programs/ProgramBlockEditView.java:274]: Good question! [Upgrade guide](https://tailwindcss.com/docs/upgrade-guide#automatic-transforms-and-filters) for Tailwind v2 -> v3 says it wasn't needed no more

File: universal-application-tool-0.0.1/app/views/applicant/ApplicantProgramSummaryView.java
Comment (syntapy) [reply universal-application-tool-0.0.1/app/views/applicant/ApplicantProgramSummaryView.java:106]: The `CONTINUE_BUTTON` is just a [reference](https://github.com/seattle-uat/civiform/blob/main/universal-application-tool-0.0.1/app/views/style/ReferenceClasses.java#L49) to `cf-continue-button`

From what I remember I couldn't figure out where the `cf-*` classes were coming from. I know they are called in `app/assets/javascript/` files, as well as through direct string literals elsewhere (i.e. not using calls to `ReferenceClasses.STYLE`)

I found some defintions for `cf-*` classes in `app/assets/stylesheets/styles.css`, but could only find definitions for classes `cf-accordion-visible`, `cf-accordion-header`, `cf-accordion-content`, and `cf-accordion-button`
cf-classes

But either way, I had just removed that and didn't find any effect on the resulting visual style, at least using the browsers / testing methods

That said, I am not certain it didn't have an adverse impact on semantics or whatnot, so maybe should review this more

File: bin/pull-image
Comment (syntapy) [reply bin/pull-image:15]: I remember at one point they changed it to not show pull progress, and then I undid that in my local changes and @bion had asked me to revert that

File: universal-application-tool-0.0.1/app/views/style/README.md
Comment (syntapy) [reply universal-application-tool-0.0.1/app/views/style/README.md:45]: No. I forgot to fix this

File: universal-application-tool-0.0.1/app/views/style/Styles.java
Comment (syntapy) [reply universal-application-tool-0.0.1/app/views/style/Styles.java:2516]: This is from the [upgrade guide](https://tailwindcss.com/docs/upgrade-guide#overflow-clip-ellipsis) for upgrading TailwindCSS v2 -> v3. Does that make a difference?

File: universal-application-tool-0.0.1/package-lock.json
Comment (syntapy) [reply universal-application-tool-0.0.1/package-lock.json:1]: Yes! I will gladly do so. I've always hated the rebase pain this file causes

File: bin/pull-image
Comment (kmcnellis) [reply bin/pull-image:15]: Ah, gotcha. could just leave a comment about the flag in the docstring above, then

File: universal-application-tool-0.0.1/project/TailwindBuilder.scala
Comment (kmcnellis) [inline universal-application-tool-0.0.1/project/TailwindBuilder.scala:17]: When I run this on my local command line, it says `Warning: Task "dist" not found` (though it doesn't seem to be doing it in the container).   Where is the dist command defined?

File: universal-application-tool-0.0.1/app/views/style/Styles.java
Comment (syntapy) [reply universal-application-tool-0.0.1/app/views/style/Styles.java:2516]: Sorry! I see what you mean now. Yeah those renamed styles aren't used anywhere so putting this in a different PR would be better

Yeah theres waay too much stuff in this PR right now!

File: universal-application-tool-0.0.1/project/TailwindBuilder.scala
Comment (syntapy) [reply universal-application-tool-0.0.1/project/TailwindBuilder.scala:17]: Oh! That means its working for you

None of that code is running (as far as I can tell?) on my system since the docker dev upgrade stuff

But I also have a non-standard system: Fedora running in QubesOS (i.e. the Xen hypervisor)

Can you run `./bin/grunt-healthcheck`? If it runs silently that means grunt is working, otherwise it would raise an error

Comment (syntapy) [reply universal-application-tool-0.0.1/project/TailwindBuilder.scala:17]: I think I need to add the `grunt-dist` package to solve that error, but why is that command needed anyways

I merely saw `grunt dist` in an online example

Comment (syntapy) [reply universal-application-tool-0.0.1/project/TailwindBuilder.scala:17]: This is moot cuz I'm doing the sbt approach

**Pull request closed without merging**
